File: Cpp/quick_sort_lomuto.cpp
#include "sort_algorithms.h"
#include <vector>
#include <algorithm>

int partition_Lomuto(std::vector<int> &array, int low, int high)
{
    int pivot = array[high];
    int i = low;
    int j = low;
    while (j < high)
    {
        if (array[j] <= pivot)
        {
            std::swap(array[i], array[j]);
            i += 1;
        }
        j += 1;
    }
    std::swap(array[i], array[high]);
    return i;
}

void quick_sort_Lomuto_helper(std::vector<int> &items, int low, int high)
{
    if (low < high)
    {
        int split_point = partition_Lomuto(items, low, high);
        quick_sort_Lomuto_helper(items, low, split_point - 1);
        quick_sort_Lomuto_helper(items, split_point + 1, high);
    }
}

std::vector<int> quick_sort_Lomuto(std::vector<int> array)
{
    quick_sort_Lomuto_helper(array, 0, array.size() - 1);
    return array;
}

File: Python/quick_sort_lomuto.py
def partition_lomuto(arr, low, high):
    pivot = arr[high]
    i = low - 1
    for j in range(low, high):
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1


def quick_sort_lomuto(arr, low, high):
    if low < high:
        pi = partition_lomuto(arr, low, high)
        quick_sort_lomuto(arr, low, pi - 1)
        quick_sort_lomuto(arr, pi + 1, high)
    return arr
